arch/risc-v/src/mpfs/mpfs_corepwm.c

```diff
@@ -0,0 +1,688 @@
+
+  /* Set the channels duty cycle by writing to the NEG_EDGE register
+    * for this channel
+    */
+  const int neg_edge_reg_offset =
+    MPFS_COREPWM_PWM1_NEG_EDGE_OFFSET +
+    (MPFS_COREPWM_PWM2_NEG_EDGE_OFFSET -
+     MPFS_COREPWM_PWM1_NEG_EDGE_OFFSET) * (channel - 1);
+
+  pwm_putreg(priv, neg_edge_reg_offset, neg_edge);
+
+  return OK;
+}
+
+
+/****************************************************************************
+ * Name: pwm_setup
+ *
+ * Description:
+ *   This method is called when the driver is opened.  The lower half driver
+ *   should configure and initialize the device so that it is ready for use.
+ *   It should not, however, output pulses until the start method is called.
+ *
+ * Input Parameters:
+ *   dev - A reference to the lower half PWM driver state structure
+ *
+ * Returned Value:
+ *   Zero on success; a negated errno value on failure
+ *
+ * Assumptions:
+ *
+ * Note:
+ *   On a MPFS CorePWM block no setting up is needed
+ *
+ ****************************************************************************/
+
+static int pwm_setup(FAR struct pwm_lowerhalf_s *dev)
+{
+  FAR struct mpfs_pwmtimer_s *priv = (FAR struct mpfs_pwmtimer_s *)dev;
+
+  pwminfo("PWMID%u\n", priv->pwmid);
+  pwm_dumpregs(priv, "Initially");
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: pwm_shutdown
+ *
+ * Description:
+ *   This method is called when the driver is closed.  The lower half driver
+ *   stop pulsed output, free any resources, disable the timer hardware, and
+ *   put the system into the lowest possible power usage state
+ *
+ * Input Parameters:
+ *   dev - A reference to the lower half PWM driver state structure
+ *
+ * Returned Value:
+ *   Zero on success; a negated errno value on failure
+ *
+ ****************************************************************************/
+
+static int pwm_shutdown(FAR struct pwm_lowerhalf_s *dev)
+{
+  FAR struct mpfs_pwmtimer_s *priv = (FAR struct mpfs_pwmtimer_s *)dev;
+
+  pwminfo("PWM%u\n", priv->pwmid);
+
+  /* Make sure that the output has been stopped */
+
+  pwm_stop(dev);
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: pwm_start
+ *
+ * Description:
+ *   (Re-)initialize the timer resources and start the pulsed output
+ *
+ * Input Parameters:
+ *   dev - A reference to the lower half PWM driver state structure
+ *   info - A reference to the characteristics of the pulsed output
+ *
+ * Returned Value:
+ *   Zero on success; a negated errno value on failure
+ *
+ ****************************************************************************/
+
+static int pwm_start(FAR struct pwm_lowerhalf_s *dev,
+                     FAR const struct pwm_info_s *info)
+{
+  int ret = OK;
+  FAR struct mpfs_pwmtimer_s *priv = (FAR struct mpfs_pwmtimer_s *)dev;
+
+  /* if frequency has not changed we just update duty */
+  if (info->frequency == priv->frequency)
```

Comment:
Is there a reason for not just setting both frequency and duty cycle always at pwm_start?